File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        let res = []

        const inorder = (root) => {
            if (root === null) return root

            inorder(root.left)
            res.push(root.val)
            inorder(root.right)
        }

        inorder(root)
        return res[k - 1]

        // Iterative:
        // let stack = []
        // let curr = root
        // let count = 0

        // while (curr || stack.length > 0) {
        //     while (curr) {
        //     stack.push(curr)
        //     curr = curr.left
        //     }

        //     curr = stack.pop()
        //     count++
        //     if (count === k) {
        //     return curr.val
        //     }
        //     curr = curr.right
        // }
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxDiameter = 0

        def dfs(node):
            nonlocal maxDiameter

            if node is None:
                return 0
                
            left = dfs(node.left)
            right = dfs(node.right)
            maxDiameter = max(maxDiameter, left + right)
            
            return 1 + max(dfs(node.left), dfs(node.right))
        
        dfs(root)
        return maxDiameter